File: python3/misc.py
# Miscellaneous utilities.

def reverse_number(n, base=10):
    result = 0
    while n > 0:
        result = result * base + n % base
        n //= base
    return result

def is_palindrome(n, base=10):
    return n == reverse_number(n, base)

def isqrt(x):
    if x < 0:
        raise ValueError('square root not defined for negative numbers')
    n = int(x)
    if n == 0:
        return 0
    a, b = divmod(n.bit_length(), 2)
    x = 2**(a+b)
    while True:
        y = (x + n//x)//2
        if y >= x:
            return x
        x = y

File: swift/misc.swift
// Miscellaneous utilities.

// TODO: Figure out how to make this generic.

func is_palindrome(number: Int64, base: Int64) -> Bool {
    return number == reverse_number(number, base: base)
}

func reverse_number(number: Int64, base: Int64) -> Int64 {
    var n = number
    var result: Int64 = 0
    while n > 0 {
        result = result * base + n % base
        n /= base
    }
    return result
}

// print(is_palindrome(123454321, base: 10))
